# Ancestral Recombination Graph for Chromosome 2
## Generation of input data
A custom perl script, [Generate_ARG_sites.pl](/scripts/Generate_ARG_sites.pl), was used to create a .sites file that represents the intersection of Chromosome 2 SNP calls generated by GATK (SmartSNPs-filtered) and iSNPcaller for a representative set of PoT/PoL isolates and candidate donors. To accelerate the analysis, the data were compressed by including only every 5th variant position. The resulting dataset, [Chr2ARGvalidSkip4.sites](/data/Chr2ARGvalidSkip4.sites), contains 7,593 variant positions.
## Running ARGweaver
ARGweaver (version X.X) was run using the following parameters:
```bash
arg-sample -s Chr2ARGvalidSkip4.sites --ntimes 100  -n 200 --sample-step 1 -m 1.2e-7 -r 1e-9 -c 20 -o Chr2ARGvalidSkip4 --overwrite
```
## Generation of a tree sequence for select target regions on chromosome 2
The above script results in the generation of 200 trees for each chromosome block that is separated from the next by a recombination event. A tree sequence was generated by extracting the consensus trees for the nine chromosome regions that span positions 500 kb, 1 Mb, 2 Mb, 3 Mb, 4 Mb, 5 Mb, 6 Mb, 7 Mb and 7.8 Mb. This was accomplishd using ARGweaver.py [ARGweaver.py](/scripts/ARGweaver.py) - a modified version of a script from the Espalier module (https://github.com/davidrasm/Espalier):
```bash
mkdir SMC_trees
mv Chr2Chr2ARGvalidSkip4*smc.gz SMC_trees/
python ARGweaver.py SMC_trees > Chr2ARGtrees.txt
```
## Creation of a TangleGram representing the tree sequence
1. Generate a lookup file listing indexes for each strain:
```bash
for f in {1..21}; do zgrep '' SMC_trees/Chr2ARGvalidSkip4.0.smc.gz | head -1 | awk -v var="$f" '{print $(var+1) "\t" var-1}' >> ARGstrain_lookup.txt; done
```
2. The TangleGram.py script was then used to create a tanglegram for the chromosome 2 tree sequence:
```bash
python TangleGram.py ARGstrain_lookup.txt ConsensusTrees.tre
```
## Extracting branch length distributions for the PoT/PoL clades
Patristic distances between every taxon pair were extracted for each of the 200 MCMC trees using [tipDistances_multi.py](/scripts/tipDistances_multi.py):
```bash
python tipDistances_multi.py SMC_trees > Chr2ARGdistances.txt
```
Next, the distances between each PoT/PoL isolate and its nearest donor were extracted using the [tipDistancesPoTPoL.pl](/scripts/tipDistancesPoTPoL.pl) script:
```bash
perl tipDistancesPoTPoL.pl Chr2ARGdistances.txt > Chr2ARGPoTPoLdistances.txt
```
## Statistics for branch length distributions
Branch length means and standard deviations were calculated using the [BranchLengths.R](/scripts/BranchLengths.R) script.

