# Molecular Dating of Wheat Blast Evolution
The validated SNP dataset was converted into a sequence alignment file, and dates of sample collection were added using the ([Create_fasta_alignment.pl](/Fig5/scripts/Create_fasta_alignment.pl) script (note: this uses alignment strings data to ensure that each site was called in every genome. These data are too large to be included in this repository and should be generated *de novo* from BLAST results using the [Create_align_strings.pl](/scripts/Create_align_strings.pl) script):
```bash
perl Create_fasta_alignment.pl Chr1Chr2Chr5_final.txt ALIGNSTRINGS
```
This produced the alignment file: Chr1Chr2Chr5_final.fasta.

## Measurement of Phylogenetic Signal
The [Chr1Chr2Chr5_undated.fasta](/Fig5/data/Chr1Chr2Chr5_undated.fasta) file comprising all SNPs on the co-inherited regions of chromosomes 1, 2 and 5 was used to build a distance tree with RAxML-NG under the GTRgamma model and 1,000 bootstrap replications. The resulting tree [Chr1Chr2Chr5_undated.fasta.bestTree.raxml](/Fig5/data/Chr1Chr2Chr5_undated.fasta.bestTree.raxml) was then used as an input to the R script [Fig5AB_PhyloSignal.pl](/Fig5/scripts/Fig5AB_PhyloSignal.R) to calculate Pearson's correlation between pylogenetic distance and sampling date, and to assess the robustness of this signal by resampling the data. 

![Fig5AB.tiff](/Fig5/Fig5AB.tiff)
Figure 5A&B. Correlation between Patristic Distance and Sampling Date.
## Choosing a Nucleotide Substitution Model
A carefully curated SNP dataset was generated by [filtering out false SNP calls](/VariantCalling.md#filtering-to-remove-false-snp-calls) and a custom script [ConstantSites.pl](/Fig5/scripts/ConstantSites.pl) was then used to create a dataset comprising all of the invariant nucleotides from the regions of the B71 reference genome used for SNP calling. Sites were included only if they were present in every samples and did NOT reside in repetitive regions of the genome. The combined variant and invariant sites data were exported in phylip format and used as input to [PartitionFinder2](https://github.com/brettc/partitionfinder). The configuration file used for the run was [partition_finder.cfg](/Fig5/data/partition_finder.cfg) and the model that was selected was HKY + G + I ([best scheme](/Fig5/data/best_scheme.txt)).
## Molecular Dating
The [Chr1Chr2Chr5_dated.fasta](/Fig5/data/Chr1Chr2Chr5_dated.fasta) file was imported into BEAUTi and the HKY + G model was selected with an invariant (+I) proportion of 0.985. Runs were performed using both the strict and relaxed clock models (to account for the different substitution rates inferred from the TempEst analysis, Figure 1), with an lower substitution rate limit of 5E-8 and an upper limit of 5E-7. After the xml file was generated upon saving, we manually edited the file to add a constantSiteWeights parameter (A:840276 C: 914905 G: 910177 T: 839104). These values were determined using the [ConstantSites.pl](/Fig5/scripts/ConstantSites.pl) script. A total 100 million iterations with 10 million burn-in samples were performed for each condition, after which the operator switches were tuned to the settings recommended at runtime completion. This step was repeated until no further recommendations were given, or until the recommended setting simply "flip-flopped" backwards and forwards after each adjustment. Once the optimal paramters had been established, we performed 20 runs for both the strict and relaxed clock settings, with each run comprising 100 million iterations, after 10 million burn-in samplings. Upon completion we combined the logs and the trees for the 5 best runs for both the strict and relaxed clock models (based on the highest ESS values for clock rate and tree height). We then used TreeAnnotator to select the maximum clade credibility tree. The final trees were drawn in FigTree v1.4.4 (http://tree.bio.ed.ac.uk/software/figtree/) and exported in .svg format and then manually edited in Inkscape (https://inkscape.org).

![BEASTtree.tiff](/Fig5/BEASTtree.tiff)

Figure 5C. Molecular Dating of Wheat Blast Evolution
